src/main/scala/za/co/absa/commons/reflect/ReflectionUtils.scala

```diff
@@ -94,47 +93,123 @@ object ReflectionUtils {
   }
 
   /**
-    * Extract a value from a given field reflexively.
-    * The field is lookup recursively in the target object class hierarchy.
+    * Extract a value from a given field regardless of its visibility.
+    * This method utilizes a mix of Java and Scala reflection mechanisms,
+    * and can extract from a compiler generated fields as well.
+    * Note: if the field has an associated Scala accessor one will be called.
+    * Consequently if the filed is lazy it will be initialized.
     *
     * @param o         target object
     * @param fieldName field name to extract value from
-    * @tparam T expected type of the field value to return
+    * @tparam A type in which the given field is declared
+    * @tparam B expected type of the field value to return
     * @return a field value
     */
-  def extractFieldValue[T](o: AnyRef, fieldName: String): T = {
-    @tailrec def findField(c: Class[_]): Field =
-      try c.getDeclaredField(fieldName)
-      catch {
-        case e: NoSuchFieldException =>
+  def extractFieldValue[A: ClassTag, B](o: AnyRef, fieldName: String): B = {
+    @tailrec
+    def reflectClassHierarchy(c: Class[_]): Option[_] =
+      if (c == classOf[AnyRef]) None
+      else {
+        val maybeValue: Option[Any] = reflectClass(c)
+        if (maybeValue.isDefined) maybeValue
```

Comment:
wouln't `getOrElse` be more readable?

Reply:
It would, but I wanted it to be tail recursive